### ISSUE TITLE
[BUG] MatX fails to build with -DMATX_BUILD_32_BIT=ON

### ISSUE THREAD

**Issue by dagardner-nv** [NVIDIA/MatX#386]
**Describe the bug**
MatX fails to build with `-DMATX_BUILD_32_BIT=ON`

**To Reproduce**
```
git clone git@github.com:NVIDIA/MatX.git matx_32
cd matx_32
mamba create -n matx_32 Python=3.8 numpy
conda activate matx_32
mkdir build && cd build
cmake -DMATX_BUILD_32_BIT=ON -DMATX_BUILD_BENCHMARKS=OFF -DMATX_BUILD_DOCS=OFF -DMATX_BUILD_EXAMPLES=ON -DMATX_BUILD_TESTS=ON ..
make -j
```

However the above steps build with `-DMATX_BUILD_32_BIT=OFF`:
```
git clone git@github.com:NVIDIA/MatX.git matx_64
cd matx_64/
mamba create -n matx_64 Python=3.8 numpy
conda activate matx_64
mkdir build && cd build
cmake -DMATX_BUILD_32_BIT=OFF -DMATX_BUILD_BENCHMARKS=OFF -DMATX_BUILD_DOCS=OFF -DMATX_BUILD_EXAMPLES=ON -DMATX_BUILD_TESTS=ON ..
make -j
```

**Expected behavior**
Builds

**Code snippers**
```
$ make -j
[  1%] Building CUDA object examples/CMakeFiles/simple_radar_pipeline.dir/simple_radar_pipeline.cu.o
[  2%] Building CUDA object examples/CMakeFiles/recursive_filter.dir/recursive_filter.cu.o
[  4%] Building CUDA object examples/CMakeFiles/spectrogram_graph.dir/spectrogram_graph.cu.o
[  5%] Building CUDA object examples/CMakeFiles/resample.dir/resample.cu.o
[  8%] Building CUDA object examples/CMakeFiles/cgsolve.dir/cgsolve.cu.o
[ 10%] Building CUDA object examples/CMakeFiles/convolution.dir/convolution.cu.o
[ 10%] Building CUDA object examples/CMakeFiles/fft_conv.dir/fft_conv.cu.o
[ 11%] Building CXX object _deps/gtest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[ 14%] Building CUDA object examples/CMakeFiles/conv2d.dir/conv2d.cu.o
[ 14%] Building CUDA object examples/CMakeFiles/black_scholes.dir/black_scholes.cu.o
[ 16%] Building CUDA object examples/CMakeFiles/spherical_harmonics.dir/spherical_harmonics.cu.o
[ 17%] Building CUDA object examples/CMakeFiles/spectrogram.dir/spectrogram.cu.o
[ 19%] Building CUDA object examples/CMakeFiles/svd_power.dir/svd_power.cu.o
[ 22%] Building CUDA object examples/CMakeFiles/mvdr_beamformer.dir/mvdr_beamformer.cu.o
[ 22%] Building CUDA object examples/CMakeFiles/qr.dir/qr.cu.o
/home/dagardner/work/tmp/test/matx_32/include/matx/core/tensor_desc.h(251): error: no operator "=" matches these operands
            operand types are: std::array<long long, 2UL> = std::array<int, 2UL>
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeContainer, StrideContainer, RANK>::InitFromShape(S2 &&) [with ShapeContainer=std::array<long long, 2UL>, StrideContainer=std::array<long long, 2UL>, RANK=2, S2=std::array<int, 2UL>]" 
(150): here
            instantiation of "matx::tensor_desc_t<ShapeContainer, StrideContainer, RANK>::tensor_desc_t(const matx::index_t (&)[M]) [with ShapeContainer=std::array<long long, 2UL>, StrideContainer=std::array<long long, 2UL>, RANK=2, M=2]" 
/home/dagardner/work/tmp/test/matx_32/include/matx/core/make_tensor.h(54): here

/home/dagardner/work/tmp/test/matx_32/include/matx/core/tensor_desc.h(251): error: no operator "=" matches these operands
            operand types are: std::array<long long, 1UL> = std::array<matx::index_t, 1UL>
          detected during:
            instantiation of "void matx::tensor_desc_t<ShapeContainer, StrideContainer, RANK>::InitFromShape(S2 &&) [with ShapeContainer=std::array<long long, 1UL>, StrideContainer=std::array<long long, 1UL>, RANK=1, S2=std::array<matx::index_t, 1UL>]" 
(150): here
            instantiation of "matx::tensor_desc_t<ShapeContainer, StrideContainer, RANK>::tensor_desc_t(const matx::index_t (&)[M]) [with ShapeContainer=std::array<long long, 1UL>, StrideContainer=std::array<long long, 1UL>, RANK=1, M=1]" 
<.......... SNIP .............>
8 errors detected in the compilation of "/home/dagardner/work/tmp/test/matx_32/examples/simple_radar_pipeline.cu".
make[2]: *** [examples/CMakeFiles/simple_radar_pipeline.dir/build.make:77: examples/CMakeFiles/simple_radar_pipeline.dir/simple_radar_pipeline.cu.o] Error 255
make[1]: *** [CMakeFiles/Makefile2:233: examples/CMakeFiles/simple_radar_pipeline.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
I had to truncate the build errors as I hit some max length in Github.

**System details (please complete the following information):**
 - OS: Debian 11.6
 - CUDA version: 11.5
 - g++ version: 10.2.1

**Additional context**
Trying to perform a 32bit build as this would help with interoperability with CuDF's `cudf::size_type` which is an `int32_t`.

**Comment by luitjens** [comment]
Thank you for reporting this.  @cliffburdick  it looks like we need use the 32 bit macro to set the default descriptor appropriately.  Right now it always sets it to 64  bit (line 471).

https://github.com/NVIDIA/MatX/blob/main/include/matx/core/tensor_desc.h#L444-L471

**Comment by cliffburdick** [comment]
Hi @dagardner-nv , 32-bit mode hasn't been tested in a very long time, and there are probably many places that would require fixing. The original purpose of it was to reduce the size of each operator so you could have very long expressions that fori under the 4K parameter limit. Are you using it for this reason, or is it purely for a slight improvement in indexing speed? We should probably get rid of it unless it's actually needed.

**Comment by cliffburdick** [comment]
Fixed in #388
